bigquery_etl/query_scheduling/dag.py

```diff
@@ -24,19 +44,93 @@ class InvalidDag(Exception):
+
+    owner: str = attr.ib()
+    start_date: str = attr.ib()
+    email: List[str] = attr.ib([])
+    depends_on_past: bool = attr.ib(False)
+    retry_delay: str = attr.ib("30m")
+    email_on_failure: bool = attr.ib(True)
+    email_on_retry: bool = attr.ib(True)
+    retries: int = attr.ib(2)
+
+    @owner.validator
+    def validate_owner(self, attribute, value):
+        """Check that owner is a valid email address."""
+        if not is_email(value):
+            raise ValueError(f"Invalid email for DAG owner: {value}.")
+
+    @email.validator
+    def validate_email(self, attribute, value):
+        """Check that provided email addresses are valid."""
+        if not all(map(lambda e: is_email(e), value)):
+            raise ValueError(f"Invalid email in DAG email: {value}.")
+
+    @retry_delay.validator
+    def validate_retry_delay(self, attribute, value):
+        """Check that retry_delay is in a valid timedelta format."""
+        if not is_timedelta_string(value):
+            raise ValueError(
+                f"Invalid timedelta definition for {attribute}: {value}."
+                "Timedeltas should be specified like: 1h, 30m, 1h15m, 1d4h45m, ..."
+            )
+
+    @start_date.validator
+    def validate_start_date(self, attribute, value):
+        """Check that start_date has YYYY-MM-DD format."""
+        if value is not None and not is_date_string(value):
+            raise ValueError(
+                f"Invalid date definition for {attribute}: {value}."
+                "Dates should be specified as YYYY-MM-DD."
+            )
+
+    def to_dict(self):
+        """Return class as a dict."""
+        return self.__dict__
+
+
+@attr.s(auto_attribs=True)
 class Dag:
     """Representation of a DAG configuration."""
 
-    def __init__(self, name, schedule_interval, default_args):
-        """Instantiate new DAG representation."""
-        self.name = name
-        self.schedule_interval = schedule_interval
-        self.default_args = default_args
-        self.tasks = []
+    name: str = attr.ib()
+    schedule_interval: str = attr.ib()
+    default_args: DagDefaultArgs
+    tasks: List[Task] = attr.ib([])
+
+    @name.validator
+    def validate_dag_name(self, attribute, value):
+        """Validate the DAG name."""
+        dag_name_pattern = re.compile("^bqetl_.+$")
+        if not dag_name_pattern.match(value):
+            raise ValueError(
+                f"Invalid DAG name {value}. Name must start with 'bqetl_'."
+            )
+
+    @schedule_interval.validator
+    def validate_schedule_interval(self, attribute, value):
+        """
+        Validate the schedule_interval format.
+
+        Schedule intervals can be either in CRON format or one of:
+        @once, @hourly, @daily, @weekly, @monthly, @yearly
+        or a timedelta []d[]h[]m
+        """
+        # https://stackoverflow.com/questions/14203122/create-a-regular-expression-for-cron-statement
+        pattern = re.compile(
```

Comment:
Can this be defined as a constant so that it doesn't have to be recomputed for each validation?